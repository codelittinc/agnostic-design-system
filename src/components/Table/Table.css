@import '../../styles/theme/index.css';
:root {
  --table-background-color: var(--structure-color-100);
  --table-header-background-color: var(--structure-color-200);
  --table-body-background-color: var(--structure-color-100);
  --table-font-family: 'Inter';
  --table-border: 1px solid var(--structure-color-300);
  --table-border-bottom: 1px solid var(--structure-color-300);
  --table-border-left: none;
  --table-border-right: none;
  --table-border-top: 1px solid var(--structure-color-300);
  --table-width: 100%;
  --table-color: var(--structure-color-800);
  --table-header-color: var(--structure-color-900);
  --table-font-family: 'Inter';
  --table-font-weight: 100;
  --table-header-font-weight: 300;
  --table-padding: 10px;
}
.table-container {
  background-color: var(--table-background-color);
  border-bottom: var(--table-border-bottom);
  border-left: var(--table-border-left);
  border-radius: var(--table-border-radius, var(--round-m));
  border-right: var(--table-border-right);
  border-top: var(--table-border-top);
  width: var(--table-width);
  color: var(--table-color);
  font-family: var(--table-font-family);
}

.table-container th {
  background-color: var(--table-header-background-color);
  border-bottom: var(--table-border-bottom);
  border-left: var(--table-border-left);
  border-right: var(--table-border-right);
  border-top: var(--table-border-top);
  color: var(--table-header-color);
  font-weight: var(--table-header-font-weight);
  padding: var(--table-padding);
}

.table-container td {
  background-color: var(--table-header-background-color);
  border-bottom: var(--table-border-bottom);
  border-left: var(--table-border-left);
  border-right: var(--table-border-right);
  border-top: var(--table-border-top);
  font-weight: var(--table-font-weight);
  padding: var(--table-padding);
}

.clickable-row {
  cursor: var(--table-clickable-row-cursor, pointer);
}
