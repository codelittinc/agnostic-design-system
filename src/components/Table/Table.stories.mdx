# Table Component

## Type: Simple

## Table Props

|        Prop        |                   Type                   |                                   Description                                    | Default | Mandatory |
| :----------------: | :--------------------------------------: | :------------------------------------------------------------------------------: | :-----: | :-------: |
|     headerList     |     array of strings or react nodes      |            Specifies the list to be used to populate the table header            |    -    |   true    |
|     footerList     |     array of strings or react nodes      |            Specifies the list to be used to populate the table footer            |    -    |   false   |
|    renderFooter    |                 function                 |                      Returns the full footer HTML structure                      |    -    |   false   |
|    tableContent    | array of array of strings or react nodes | Each array inside the matrix represents a row in the table with all its content. |    -    |   true    |
| variablesClassName |                  string                  |             Specifies the custom class to apply styles to the table              |    -    |   false   |

## Getting Started

```javascript
import { Table } from '@codelittinc/agnostic-design-system';

const props = {
  headerList: ['First Name', 'Last Name', 'Email'],
  tableContent: [
    ['Ana', 'Robin', 'ana@mail.com'],
    ['Beth', 'Julius', 'beth@mail.com'],
    ['Ruth', 'Katherine', 'ruth@mail.com'],
    ['Julia', 'Mark', 'julia@mail.com'],
    ['Nath', 'Sloan', 'nath@mail.com']
  ]
};

const Table = props => {
  return (
    <Table
      headerList={props.headerList}
      tableContent={props.tableContent}
      variablesClassName='my-custom-table'
    />
  );
};

export default Table;
```

## CSS Variables Available

| Property |    Attribute     | State |
| :------: | :--------------: | :---: |
|  table   | background-color |       |
|  table   |  border-bottom   |       |
|  table   |   border-left    |       |
|  table   |   border-right   |       |
|  table   |    border-top    |       |
|  table   |      width       |       |
|  table   |      color       |       |
|  table   |   font-family    |       |
| table th | background-color |       |
| table th |  border-bottom   |       |
| table th |   border-left    |       |
| table th |   border-right   |       |
| table th |    border-top    |       |
| table th |      color       |       |
| table th |   font-weight    |       |
| table th |     padding      |       |
| table th | background-color |       |
| table th |  border-bottom   |       |
| table th |   border-left    |       |
| table th |   border-right   |       |
| table th |    border-top    |       |
| table th |   font-weight    |       |
| table th |     padding      |       |
