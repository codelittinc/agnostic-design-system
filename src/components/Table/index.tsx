import React, { useEffect, useState } from 'react';
import ArrowDown from '@/assets/images/icons/web/arrow-down.svg';
import ArrowUp from '@/assets/images/icons/web/arrow-up.svg';
import classNames from 'classnames';

import styles from '@/components/Table/Table.css';
export interface Props {
  footerList?: React.ReactNode[];
  headerList: React.ReactNode[] | string[];
  onRowClick?: (row: React.ReactNode, rowIndex: number) => void;
  onRowMouseEnter?: (row: React.ReactNode, rowIndex: number) => void;
  onRowMouseLeave?: (row?: React.ReactNode, rowIndex?: number) => void;
  renderFooter?: () => React.ReactNode;
  sortIcon?: React.ReactNode | { iconAsc: React.ReactNode; iconDesc: React.ReactNode };
  setSortState?: (value: {
    sortField: string | React.ReactNode;
    sortOrder: 'ASC' | 'DESC';
  }) => void;
  tableContent: React.ReactNode[][];
  variablesClassName?: string;
}

const Table = (props: Props) => {
  const {
    footerList,
    headerList,
    onRowClick,
    onRowMouseEnter,
    onRowMouseLeave,
    renderFooter,
    sortIcon,
    setSortState,
    tableContent,
    variablesClassName
  } = props;

  const [sortStatus, setSortStatus] = useState<{
    sortField: string | React.ReactNode;
    sortOrder: 'ASC' | 'DESC';
  }>({
    sortField: headerList[0],
    sortOrder: 'ASC'
  });

  useEffect(() => {
    setSortState && setSortState(sortStatus);
  }, [sortStatus, setSortState]);

  const handleRowClick = (row, rowIndex) => {
    if (onRowClick) {
      onRowClick(row, rowIndex);
    }
  };

  const handleRowMouseEnter = (row, rowIndex) => {
    if (onRowMouseEnter) {
      onRowMouseEnter(row, rowIndex);
    }
  };

  const handleRowMouseLeave = (row, rowIndex) => {
    if (onRowMouseLeave) {
      onRowMouseLeave(row, rowIndex);
    }
  };

  const updateSortingStatus = header => {
    if (setSortState && typeof header === 'string') {
      setSortStatus({
        sortField: header,
        sortOrder:
          sortStatus.sortField === header && sortStatus.sortOrder === 'ASC' ? 'DESC' : 'ASC'
      });
    }
  };

  const renderSortIcon = (header: string) => {
    const icons =
      !!setSortState && !sortIcon
        ? {
            iconAsc: <ArrowUp className={classNames(styles['arrow-icon'])} />,
            iconDesc: <ArrowDown className={classNames(styles['arrow-icon'])} />
          }
        : sortIcon;

    if (!!icons && typeof icons === 'object' && 'iconAsc' in icons && 'iconDesc' in icons) {
      return sortStatus?.sortOrder === 'ASC' && sortStatus.sortField === header
        ? icons.iconAsc
        : icons.iconDesc;
    }

    return icons;
  };

  return (
    <table
      className={classNames(variablesClassName, styles['table-container'])}
      data-testid='table-element'
    >
      <thead data-testid='table-header'>
        <tr>
          {headerList.map((header, headerIndex) => (
            <th
              onClick={() => updateSortingStatus(header)}
              key={headerIndex}
              className={setSortState && classNames(styles['sortable-header'])}
            >
              {header}
              {renderSortIcon(header)}
            </th>
          ))}
        </tr>
      </thead>
      <tbody data-testid='table-body'>
        {tableContent.map((row, rowIndex) => (
          <tr
            className={onRowClick ? styles['clickable-row'] : styles['non-clickable-row']}
            tabIndex={0}
            key={rowIndex}
            onClick={() => handleRowClick(row, rowIndex)}
            onMouseEnter={() => handleRowMouseEnter(row, rowIndex)}
            onMouseLeave={() => handleRowMouseLeave(row, rowIndex)}
            onKeyPress={event => {
              if (event.key === 'Enter') {
                handleRowClick(row, rowIndex);
              }
            }}
          >
            {row.map((content, contentIndex) => (
              <td key={`${rowIndex} - ${contentIndex}`}> {content}</td>
            ))}
          </tr>
        ))}
      </tbody>
      {renderFooter
        ? renderFooter()
        : footerList && (
            <tfoot>
              <tr>
                {footerList.map((foot, footIndex) => (
                  <th key={footIndex}>{foot}</th>
                ))}
              </tr>
            </tfoot>
          )}
    </table>
  );
};

export default Table;

Table.defaultProps = {
  headerList: [],
  tableContent: [[]]
};
