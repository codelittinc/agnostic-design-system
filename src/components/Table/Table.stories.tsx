import React from 'react';
import Table from '@/components/Table';
import { Meta } from '@storybook/react';
import mdx from './Table.stories.mdx';
import classNames from 'classnames';
import styles from './Table.stories.css';
import Icon from '../../assets/images/icons/web/arrow-down-black.svg';

export default {
  title: 'Components/Table',
  component: Table,
  parameters: {
    docs: {
      page: mdx
    }
  }
} as Meta;

const Template = args => {
  return <Table {...args} />;
};

export const Default = Template.bind({});
Default.args = {
  headerList: ['First Name', 'Last Name', 'Handle', 'User Type', 'Status'],
  tableContent: [
    ['John', 'Oto', '@timo', 'type1', 'Active'],
    ['Mark', 'Howard', '@mdo', 'type2', 'Inactive'],
    ['Jacob', 'Martin', '@mdo', 'type1', 'Active'],
    ['Lary', 'King', '@timo', 'type1', 'Inactive'],
    ['Helen', 'Bonham', '@timo', 'type2', 'Active'],
    ['Felix', 'Dougherty', '@timo', 'type1', 'Active'],
    ['Joseph', 'Thomas', '@timo', 'type1', 'Inactive'],
    ['Richard', 'Cardenas', '@timo', 'type1', 'Active'],
    ['Karl', 'Roy', '@twitter', 'type2', 'Active'],
    ['Sean', 'Steppe', '@timo', 'type1', 'Inactive'],
    ['Samuel', 'Marriott', '@twitter', 'type1', 'Inactive']
  ]
};

export const TableWithIcons = Template.bind({});
TableWithIcons.args = {
  headerList: ['First Name', 'Last Name', 'Handle', 'User Type', 'Status'],
  headerButtonsList: [
    <Icon key='1' />,
    <Icon key='2' />,
    <Icon key='3' />,
    <Icon key='4' />,
    <Icon key='5' />
  ],
  tableContent: [
    ['John', 'Oto', '@timo', 'type1', 'Active'],
    ['Mark', 'Howard', '@mdo', 'type2', 'Inactive'],
    ['Jacob', 'Martin', '@mdo', 'type1', 'Active'],
    ['Lary', 'King', '@timo', 'type1', 'Inactive'],
    ['Helen', 'Bonham', '@timo', 'type2', 'Active'],
    ['Felix', 'Dougherty', '@timo', 'type1', 'Active'],
    ['Joseph', 'Thomas', '@timo', 'type1', 'Inactive'],
    ['Richard', 'Cardenas', '@timo', 'type1', 'Active'],
    ['Karl', 'Roy', '@twitter', 'type2', 'Active'],
    ['Sean', 'Steppe', '@timo', 'type1', 'Inactive'],
    ['Samuel', 'Marriott', '@twitter', 'type1', 'Inactive']
  ]
};

const firstName = <p>First Name</p>;
const lastName = <p>Last Name</p>;
const type = <p>Type</p>;
const handle = <p>Type</p>;
const status = <p>Status</p>;

export const PassingElementsAsHeader = Template.bind({});
PassingElementsAsHeader.args = {
  headerList: [firstName, lastName, handle, type, status],
  tableContent: [
    ['John', 'Oto', '@timo', 'type1', 'Active'],
    ['Mark', 'Howard', '@mdo', 'type2', 'Inactive'],
    ['Jacob', 'Martin', '@mdo', 'type1', 'Active'],
    ['Lary', 'King', '@timo', 'type1', 'Inactive'],
    ['Helen', 'Bonham', '@timo', 'type2', 'Active'],
    ['Felix', 'Dougherty', '@timo', 'type1', 'Active'],
    ['Joseph', 'Thomas', '@timo', 'type1', 'Inactive'],
    ['Richard', 'Cardenas', '@timo', 'type1', 'Active'],
    ['Karl', 'Roy', '@twitter', 'type2', 'Active'],
    ['Sean', 'Steppe', '@timo', 'type1', 'Inactive'],
    ['Samuel', 'Marriott', '@twitter', 'type1', 'Inactive']
  ]
};

export const CustomStyledTable = Template.bind({});
CustomStyledTable.args = {
  headerList: ['First Name', 'Last Name', 'Handle', 'User Type', 'Status'],
  tableContent: [
    ['John', 'Oto', '@timo', 'type1', 'Active'],
    ['Mark', 'Howard', '@mdo', 'type2', 'Inactive'],
    ['Jacob', 'Martin', '@mdo', 'type1', 'Active'],
    ['Lary', 'King', '@timo', 'type1', 'Inactive'],
    ['Helen', 'Bonham', '@timo', 'type2', 'Active'],
    ['Felix', 'Dougherty', '@timo', 'type1', 'Active'],
    ['Joseph', 'Thomas', '@timo', 'type1', 'Inactive'],
    ['Richard', 'Cardenas', '@timo', 'type1', 'Active'],
    ['Karl', 'Roy', '@twitter', 'type2', 'Active'],
    ['Sean', 'Steppe', '@timo', 'type1', 'Inactive'],
    ['Samuel', 'Marriott', '@twitter', 'type1', 'Inactive']
  ],
  variablesClassName: classNames(styles['custom-table'])
};
