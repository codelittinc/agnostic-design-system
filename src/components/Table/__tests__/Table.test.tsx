import React, { useState } from 'react';
import Table from '@/components/Table';
import { render, screen, fireEvent } from '@testing-library/react';

const headerList = ['Name', 'Month', 'Gift'];
const body = [
  ['Lisa', 'December', 'Car'],
  ['Tina', 'June', 'Dog'],
  ['Callie', 'August', 'Flowers'],
  ['Ruth', 'September', 'Diamonds']
];

describe('Table', () => {
  it('renders correctly', () => {
    const { asFragment } = render(<Table headerList={headerList} tableContent={body} />);
    expect(asFragment()).toMatchSnapshot();
  });

  describe('Table output', () => {
    it('Contains the correct number of header tags', () => {
      render(<Table headerList={headerList} tableContent={body} />);
      const header = screen.getByTestId('table-header').querySelectorAll('th');
      expect(header.length).toBe(headerList.length);
    });

    it('Displays the correct header content', () => {
      render(<Table headerList={headerList} tableContent={body} />);
      const header = screen.getByTestId('table-header');
      expect(header).toHaveTextContent('Name');
      expect(header).toHaveTextContent('Month');
      expect(header).toHaveTextContent('Gift');
    });

    it('Contains the correct number of rows', () => {
      render(<Table headerList={headerList} tableContent={body} />);
      const rows = screen.getByTestId('table-body').querySelectorAll('tr');
      expect(rows.length).toBe(body.length);
    });

    it('Displays the correct body content', () => {
      render(<Table headerList={headerList} tableContent={body} />);
      const header = screen.getByTestId('table-body');
      expect(header).toHaveTextContent('Lisa');
      expect(header).toHaveTextContent('December');
      expect(header).toHaveTextContent('Car');
    });
  });

  describe('Sortable table', () => {
    const SortableTable = () => {
      const [sortState, setSortState] = useState({});

      return (
        <div>
          <p data-testid='field'>Sort Field: {'sortField' in sortState && sortState.sortField}</p>
          <p data-testid='order'>Sort Order: {'sortOrder' in sortState && sortState.sortOrder}</p>
          <Table headerList={headerList} tableContent={body} setSortState={setSortState} />
        </div>
      );
    };

    it('Displays the correct start sorting order and field', () => {
      render(<SortableTable />);
      const sortField = screen.getByTestId('field');
      const sortOrder = screen.getByTestId('order');
      expect(sortField).toHaveTextContent('Sort Field: Name');
      expect(sortOrder).toHaveTextContent('Sort Order: ASC');
    });

    it('Displays the correct sorting field after it is updated', () => {
      render(<SortableTable />);
      fireEvent.click(screen.getByText(/Month/));
      const sortField = screen.getByTestId('field');
      expect(sortField).toHaveTextContent('Sort Field: Month');
    });

    it('Displays the correct sorting order after it is updated', () => {
      render(<SortableTable />);
      const secondHeaderElement = screen.getByText(/Month/);
      fireEvent.click(secondHeaderElement);
      fireEvent.click(secondHeaderElement);
      const sortField = screen.getByTestId('field');
      const sortOrder = screen.getByTestId('order');
      expect(sortField).toHaveTextContent('Sort Field: Month');
      expect(sortOrder).toHaveTextContent('Sort Order: DESC');
    });
  });
});
