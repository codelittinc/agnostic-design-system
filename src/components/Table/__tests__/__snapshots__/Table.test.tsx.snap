// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders correctly 1`] = `
<DocumentFragment>
  <table
    class="table-container"
    data-testid="table-element"
  >
    <thead
      data-testid="table-header"
    >
      <tr>
        <th>
          Name
        </th>
        <th>
          Month
        </th>
        <th>
          Gift
        </th>
      </tr>
    </thead>
    <tbody
      data-testid="table-body"
    >
      <tr
        class="non-clickable-row"
        tabindex="0"
      >
        <td>
           Lisa
        </td>
        <td>
           December
        </td>
        <td>
           Car
        </td>
      </tr>
      <tr
        class="non-clickable-row"
        tabindex="0"
      >
        <td>
           Tina
        </td>
        <td>
           June
        </td>
        <td>
           Dog
        </td>
      </tr>
      <tr
        class="non-clickable-row"
        tabindex="0"
      >
        <td>
           Callie
        </td>
        <td>
           August
        </td>
        <td>
           Flowers
        </td>
      </tr>
      <tr
        class="non-clickable-row"
        tabindex="0"
      >
        <td>
           Ruth
        </td>
        <td>
           September
        </td>
        <td>
           Diamonds
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
